odoo/http.py

```diff
@@ -1,1710 +1,1674 @@
 
-ODOO_DISABLE_SESSION_GC = str2bool(os.environ.get('ODOO_DISABLE_SESSION_GC', '0'))
+# =========================================================
+# Core type-specialized dispatchers
+# =========================================================
 
-if ODOO_DISABLE_SESSION_GC:
-    # empty function, in case another module would be
-    # calling it out of setup_session()
-    session_gc = lambda s: None
+_dispatchers = {}
 
+class Dispatcher(ABC):
+    routing_type: str
 
-#----------------------------------------------------------
-# WSGI Layer
-#----------------------------------------------------------
-# Add potentially missing (older ubuntu) font mime types
-mimetypes.add_type('application/font-woff', '.woff')
-mimetypes.add_type('application/vnd.ms-fontobject', '.eot')
-mimetypes.add_type('application/x-font-ttf', '.ttf')
-# Add potentially missing (detected on windows) svg mime types
-mimetypes.add_type('image/svg+xml', '.svg')
+    @classmethod
+    def __init_subclass__(cls):
+        super().__init_subclass__()
+        _dispatchers[cls.routing_type] = cls
 
-class Response(werkzeug.wrappers.Response):
-    """ Response object passed through controller route chain.
+    def __init__(self, request):
+        self.request = request
 
-    In addition to the :class:`werkzeug.wrappers.Response` parameters, this
-    class's constructor can take the following additional parameters
-    for QWeb Lazy Rendering.
+    @classmethod
+    @abstractmethod
+    def is_compatible_with(cls, request):
+        """
+        Determine if the current request is compatible with this
+        dispatcher.
+        """
 
-    :param basestring template: template to render
-    :param dict qcontext: Rendering context to use
-    :param int uid: User id to use for the ir.ui.view render call,
-                    ``None`` to use the request's user (the default)
+    def pre_dispatch(self, rule, args):
+        """
+        Prepare the system before dispatching the request to its
+        controller. This method is often overridden in ir.http to
+        extract some info from the request query-string or headers and
+        to save them in the session or in the context.
+        """
+        routing = rule.endpoint.routing
+        self.request.session.can_save = not routing.get('save_session', True)
+
+        set_header = self.request.future_response.headers.set
+        cors = routing.get('cors')
+        if cors:
+            set_header('Access-Control-Allow-Origin', cors)
+            set_header('Access-Control-Allow-Methods', (
+                'POST' if routing['type'] == 'json'
+                else ', '.join(routing['methods'] or ['GET', 'POST'])
+            ))
+
+        if cors and self.request.httprequest.method == 'OPTIONS':
+            set_header('Access-Control-Max-Age', CORS_MAX_AGE)
+            set_header('Access-Control-Allow-Headers',
+                       'Origin, X-Requested-With, Content-Type, Accept, Authorization')
```

Comment:
@Julien00859 This header will forbid custom headers in CORS requests :thinking: Why do you want to restrict this list? In the cors preflight, the browser sends headers that indicate the HTTP method and headers that will be used in the actual request. If you response with a list of hard-coded headers, it's therefore not possible to use custom headers. We should allow requested headers
```python
set_header('Access-control-Allow-Headers', self.request.httprequest.environ.get("Access-Control-Request-Headers"))
```